esmvaltool/diag_scripts/rainfarm/rainfarm.jl

```diff
@@ -0,0 +1,153 @@
+diag_base = climolist["diagnostic"]
+
+println(diag_base, ": starting routine")
+println(diag_base, ": creating work and plot directories")
+work_dir = settings["work_dir"]
+run_dir = settings["run_dir"]
+mkpath(work_dir)
+mkpath(run_dir)
+cd(run_dir)
+
+# setup provenance file and list
+provenance_file = joinpath(run_dir, "diagnostic_provenance.yml")
+provenance = Dict()
+
+# Reading parameters from the settings
+nf = get(settings, "nf", 2)
+slope = get(settings, "slope", 0.0)
+kmin = get(settings, "kmin", 1)
+nens = get(settings, "nens", 1)
+weights_climo = get(settings, "weights_climo", "")
+conserv_glob = get(settings, "conserv_glob", false)
+conserv_smooth = get(settings, "conserv_smooth", true) 
+auxiliary_data_dir = get(settings, "auxiliary_data_dir", "") 
+
+if weights_climo isa Bool # Compatibility with old standard
+    weights_climo = ""
+end
+
+# Conservation options
+if (conserv_glob)
+    println("Conserving global field")
+elseif (conserv_smooth)
+    println("Smooth conservation")
+else
+    println("Box conservation")
+end
+
+for (infile, value) in metadata
+    (infilename, ) = splitext(basename(infile))
+    outfilename = joinpath(work_dir, infilename * "_downscaled")
+
+    println(diag_base, ": calling RainFARM for ", infilename)
+
+    (pr, lon_mat, lat_mat) = read_netcdf2d(infile, varname)
+
+    # Ensure grid is square and with even dims
+    nmin = min(size(pr)[1], size(pr)[2])
+    nmin = floor(Int, nmin / 2) * 2
+    pr = pr[1:nmin, 1:nmin, :]
+    if (ndims(lon_mat) == 1)
+        lon_mat = lon_mat[1:nmin]
+        lat_mat = lat_mat[1:nmin]
+    else
+        lon_mat = lon_mat[1:nmin, 1:nmin]
+        lat_mat = lat_mat[1:nmin, 1:nmin]
+    end
+
+    (lon_f, lat_f) = lon_lat_fine(lon_mat, lat_mat, nf);
+
+    # Automatic spectral slope
+    if (slope == 0.)
+        (fxp, ftp)=fft3d(pr)
+        slope =fitslopex(fxp, kmin=kmin)
+        println("Computed spatial spectral slope: ", slope)
+    else
+        println("Fixed spatial spectral slope: ", slope)
+    end
+
+    if weights_climo != ""
+        if weights_climo[1] != '/'
+            weights_climo = joinpath(auxiliary_data_dir, weights_climo)
+        end
+        println("Using external climatology for weights: ", weights_climo)
+        fileweights = joinpath(work_dir, infilename * "_w.nc")
+
+        ww = rfweights(weights_climo, infile, nf,
+                       weightsfn = fileweights, varname = varname,
+                       fsmooth = conserv_smooth)
+    else
+       ww = 1.
+    end
+
+    for iens=1:nens
+        println("Realization ", iens)
+        rd=rainfarm(pr, slope, nf, ww, fglob = conserv_glob,
+                    fsmooth = conserv_smooth, verbose=true)
+        fname=@sprintf("%s_%04d.nc", outfilename, iens);
+        write_netcdf2d(fname, rd, lon_f, lat_f, varname, infile)
+        xprov = provenance_record(infile)
+        provenance[fname] = xprov
+     end
+end
+
+# Write provenance file
+create_yaml(provenance, provenance_file)
```

Comment:
A bit too late, but note that you could have used the json library to write yaml, since json is a subset of yaml, and the julia json library does support writing files.